File: server/src/models/person/master/individual.rs
use sea_orm::entity::prelude::*;

#[sea_orm::model]
#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
#[sea_orm(schema_name = "person_master", table_name = "individuals")]

pub struct Model {
    #[sea_orm(primary_key)]
    pub id: Uuid,
    pub code: String,
    pub name: String,
    pub front_title: Option<String>,
    pub last_title: Option<String>,
    pub birth_date: Date,
    pub birth_place: String,
    #[sea_orm(default_value = "00000000-0000-0000-0000-000000000000")]
    pub gender_id: Uuid,
    #[sea_orm(default_value = "00000000-0000-0000-0000-000000000000")]
    pub religion_id: Uuid,
    #[sea_orm(default_value = "00000000-0000-0000-0000-000000000000")]
    pub occupation_id: Uuid,
    #[sea_orm(default_value = "00000000-0000-0000-0000-000000000000")]
    pub education_id: Uuid,
    #[sea_orm(default_value = "00000000-0000-0000-0000-000000000000")]
    pub income_id: Uuid,
    #[sea_orm(default_value = "00000000-0000-0000-0000-000000000000")]
    pub identification_type_id: Uuid,
    #[sea_orm(default_value = "00000000-0000-0000-0000-000000000000")]
    pub marital_status_id: Uuid,
    #[sea_orm(default_value = "00000000-0000-0000-0000-000000000000")]
    pub profession_id: Uuid,
    #[sea_orm(default_value = "00000000-0000-0000-0000-000000000000")]
    pub age_classification_id: Uuid,
    #[sea_orm(default_value = false)]
    pub is_special_need: bool,
    #[sea_orm(default_value = false)]
    pub is_social_protection_card_recipient: bool,
    #[sea_orm(default_value = false)]
    pub is_deceased: bool,
    pub created_at: Option<DateTime>,
    pub updated_at: Option<DateTime>,
    pub deleted_at: Option<DateTimeWithTimeZone>,
    pub sync_at: Option<DateTime>,
    pub created_by: Option<Uuid>,
    pub updated_by: Option<Uuid>,
    #[sea_orm(belongs_to, from = "gender_id", to = "id")]
    pub gender: BelongsTo<crate::models::person::reference::gender::Entity>,
    #[sea_orm(belongs_to, from = "religion_id", to = "id")]
    pub religion: BelongsTo<crate::models::person::reference::religion::Entity>,
    #[sea_orm(belongs_to, from = "occupation_id", to = "id")]
    pub occupation: BelongsTo<crate::models::person::reference::occupation::Entity>,
    #[sea_orm(belongs_to, from = "income_id", to = "id")]
    pub income: BelongsTo<crate::models::person::reference::income::Entity>,
    #[sea_orm(belongs_to, from = "identification_type_id", to = "id")]
    pub identification_type: BelongsTo<crate::models::person::reference::identification_type::Entity>,
    #[sea_orm(belongs_to, from = "marital_status_id", to = "id")]
    pub marital_status: BelongsTo<crate::models::person::reference::marital_status::Entity>,
    #[sea_orm(belongs_to, from = "profession_id", to = "id")]
    pub profession: BelongsTo<crate::models::person::reference::profession::Entity>,
    #[sea_orm(belongs_to, from = "education_id", to = "id")]
    pub education: BelongsTo<crate::models::literate::educations::Entity>,
    #[sea_orm(belongs_to, from = "age_classification_id", to = "id")]
    pub age_classification: BelongsTo<crate::models::person::reference::age_classification::Entity>,
    #[sea_orm(has_one)]
    pub biodata: HasOne<super::biodata::Entity>,
    #[sea_orm(has_one)]
    pub user: HasOne<crate::models::auth::user::Entity>,
    #[sea_orm(has_one)]
    pub lecturer: HasOne<crate::models::academic::lecturer::master::lecturers::Entity>,
    #[sea_orm(has_many)]
    pub employees: HasMany<crate::models::institution::master::employees::Entity>,
    #[sea_orm(has_many)]
    pub candidates: HasMany<crate::models::academic::candidate::master::candidates::Entity>,
    #[sea_orm(has_many)]
    pub evaluators: HasMany<crate::models::academic::prior_learning_recognition::transaction::evaluators::Entity>,
    #[sea_orm(has_many)]
    pub students: HasMany<crate::models::academic::student::master::students::Entity>,
}


impl ActiveModelBehavior for ActiveModel {}

pub struct IndividualToArchive;

impl Linked for IndividualToArchive {
    type FromEntity = Entity;
    type ToEntity = crate::models::document::transaction::archives::Entity;

    fn link(&self) -> Vec<RelationDef> {
        let rel: RelationDef = crate::models::document::transaction::archives::Entity::belongs_to(Entity)
            .from(crate::models::document::transaction::archives::Column::ArchiveableId)
            .to(Column::Id)
            .on_condition(|_left, _right| {
                crate::models::document::transaction::archives::Column::ArchiveableType.eq("Individual").into()
            })
            .into();
        vec![rel.rev()]
    }
}

File: client/src/models/person/master/Individual.ts
import { PersonReferenceGender } from "../reference/Gender";
import { PersonReferenceReligion } from "../reference/Religion";
import { PersonReferenceIdentificationType } from "../reference/IdentificationType";
import { PersonReferenceIncome } from "../reference/Income";
import { PersonReferenceMaritalStatus } from "../reference/MaritalStatus";
import { PersonReferenceOccupation } from "../reference/Occupation";
import { PersonReferenceAgeClassification } from "../reference/AgeClassification";

export interface PersonMasterIndividual {
  id: string;
  code: string;
  name: string;
  front_title: string | null;
  last_title: string | null;
  birth_date: string;
  birth_place: string;
  gender_id: string;
  religion_id: string;
  occupation_id: string | null;
  education_id: string | null;
  income_id: string;
  identification_type_id: string;
  marital_status_id: string;
  profession_id: string;
  age_classification_id: string | null;
  is_special_need: boolean;
  is_social_protection_card_recipient: boolean;
  is_deceased: boolean;
  created_at: string | null;
  updated_at: string | null;
  sync_at: string | null;
  deleted_at: string | null;
  created_by: string | null;
  updated_by: string | null;
};

export const initialPersonMasterIndividual: PersonMasterIndividual = {
  id: "",
  code: "",
  name: "",
  birth_date: "",
  birth_place: "",
  gender_id: "00000000-0000-0000-0000-000000000000",
  religion_id: "00000000-0000-0000-0000-000000000000",
  occupation_id: "00000000-0000-0000-0000-000000000000",
  education_id: "00000000-0000-0000-0000-000000000000",
  income_id: "00000000-0000-0000-0000-000000000000",
  identification_type_id: "3d59fc95-b07d-46ad-95ff-206b7e7f253f",
  marital_status_id: "00000000-0000-0000-0000-000000000000",
  profession_id: "00000000-0000-0000-0000-000000000000",
  is_special_need: false,
  is_social_protection_card_recipient: false,
  is_deceased: false,
  front_title: null,
  last_title: null,
  age_classification_id: null,
  created_at: null,
  updated_at: null,
  sync_at: null,
  deleted_at: null,
  created_by: null,
  updated_by: null,
};

export interface PersonMasterIndividualDataObject {
  individual: PersonMasterIndividual;
  gender: PersonReferenceGender | null;
  religion: PersonReferenceReligion | null;
  identification_type: PersonReferenceIdentificationType | null;
  income: PersonReferenceIncome | null;
  marital_status: PersonReferenceMaritalStatus | null;
  occupation: PersonReferenceOccupation | null;
  profession: any | null; // PersonReferenceProfession
  education?: any | null; // LiterateEducation
  age_classification: PersonReferenceAgeClassification | null;
  biodata: any | null; // PersonMasterBiodata
  picture: any | null; // PersonMasterImage
  user: any | null; // AuthUser
  lecturer: any | null; // AcademicLecturerMasterLecturer
  students: any[] | null; // AcademicStudentMasterStudent[]
  employees: any[] | null; // InstitutionMasterEmployee[]
  candidates: any[] | null; // AcademicCandidateMasterCandidate[]
  evaluators: any[] | null; // AcademicPriorLearningRecognitionTransactionEvaluator[]
  family_card_members: any[] | null; // PersonMasterFamilyCardMember[]
}